// @razaman2/object-manager/dist/index.mjs
// src/ObjectManager.ts
var ObjectManager = class {
  constructor(object = {}, options = {}) {
    this.object = object;
    this.options = options;
    this.skipped = [Date];
    var _a;
    this.skipped = options.skip === false ? [] : this.skipped.concat((_a = options.skip) != null ? _a : []);
  }
  paths(object = this.object, path) {
    return Object.entries(object).reduce((paths, [key, val]) => {
      var _a, _b, _c, _d;
      const previous = { path: `${path ? `${path}.` : ""}${key}` };
      if ((_c = (_b = (_a = this.options.paths) == null ? void 0 : _a.test) == null ? void 0 : _b.call(_a, previous.path)) != null ? _c : true) {
        if (typeof val === "object") {
          try {
            if ((typeof this.options.functions === "boolean" ? this.options.functions : true) && Object.values(val).some((item) => typeof item === "function")) {
              paths.push(key);
            } else {
              const nested = { paths: this.paths(val, previous.path) };
              if (nested.paths.length === 0 || ((_d = this.options.paths) == null ? void 0 : _d.full)) {
                paths.push(key);
              }
              nested.paths.forEach((segment) => {
                paths.push(`${key}.${segment}`);
              });
            }
          } catch (e) {
            paths.push(key);
          }
        } else {
          paths.push(key);
        }
      }
      return paths;
    }, []);
  }
  get(param1, param2) {
    var _a;
    const { path, alternative, object = this.object } = typeof param1 === "string" || typeof param1 === "number" ? { path: param1, alternative: param2 } : param1 != null ? param1 : {};
    if (typeof path === "number") {
      return this.get(path.toString(), param2);
    } else if (typeof path === "string") {
      try {
        return (_a = path.split(".").reduce((object2, segment) => {
          return object2[segment];
        }, object)) != null ? _a : alternative;
      } catch (e) {
        return alternative;
      }
    } else {
      return object.hasOwnProperty("") ? object[""] : (param1 == null ? void 0 : param1.hasOwnProperty("alternative")) ? alternative : object;
    }
  }
  set(param1, param2) {
    if (arguments.length === 1) {
      if (typeof param1 === "object") {
        if (param1.hasOwnProperty("value")) {
          if (param1.hasOwnProperty("path")) {
            this.set(param1.path, param1.value);
          } else {
            this.set(param1.value);
          }
        } else if (param1.hasOwnProperty("object")) {
          this.copy(param1.object);
        } else {
          this.copy(param1);
        }
      } else {
        this.set("", param1);
      }
    } else {
      if (typeof param1 === "number") {
        this.set(param1.toString(), param2);
      } else if (typeof param1 === "string") {
        param1.split(".").reduce((object, segment, index, segments) => {
          if (index === segments.length - 1) {
            object[segment] = param2;
          } else if (object.hasOwnProperty(segment)) {
            return object[segment];
          } else {
            object[segment] = RegExp(`${segment}.\\d+(?:\\.|$)`).test(param1) ? [] : typeof param2 === "object" ? Object.setPrototypeOf({}, Object.getPrototypeOf(param2)) : {};
          }
          return object[segment];
        }, this.object);
      } else {
        throw new Error(JSON.stringify({
          code: "invalid_path",
          message: "path must be typeof string or number."
        }));
      }
    }
    return this.object;
  }
  copy(object) {
    this.object = ObjectManager.on(object).clone();
    return this;
  }
  clone(object = this.object) {
    if (this.skip(object)) {
      return object;
    } else {
      return Object.setPrototypeOf(
        Object.entries(object).reduce((object2, [key, val]) => {
          try {
            if (Array.isArray(val)) {
              object2[key] = val.map((item) => {
                if (typeof item === "object") {
                  return Object.setPrototypeOf(this.clone(item), Object.getPrototypeOf(item));
                } else {
                  return item;
                }
              });
            } else if (typeof val === "object") {
              object2[key] = Object.setPrototypeOf(this.clone(val), Object.getPrototypeOf(val));
            } else {
              object2[key] = val;
            }
          } catch (e) {
            object2[key] = val;
          }
          return object2;
        }, Array.isArray(object) ? [] : {}),
        Object.getPrototypeOf(object)
      );
    }
  }
  static on(object = {}, options = {}) {
    return new ObjectManager(object, options);
  }
  skip(object) {
    for (const item of this.skipped) {
      if (object instanceof item) {
        return true;
      }
    }
    return false;
  }
};

// src/index.ts
var src_default = ObjectManager;
export {
  src_default as default
};
